spec/execution_tests.lua

```diff
@@ -1749,6 +1749,71 @@ function execution_tests.run(compile_file, backend, _ENV, only_compile)
         end)
     end)
 
+    describe("For-in loops", function()
+        compile([[
+            local function iter(state: any, current: any): (any, any)
+                if current as integer == #(state as {any}) then
+                    return nil, nil
+                end
+
+                return current as integer + 1, (state as {any})[current as integer + 1]
+            end
+
+            typealias iterfn = (any, any) -> (any, any)
+            local function xpairs(xs: {any}): (iterfn, any, any)
+                return iter, xs, 0
+            end
+
+
+            export function double(xs: {integer}): {integer}
+                local out: {integer} = {}
+                for i, x in xpairs(xs) do
+                    out[#out + 1] = x as integer * 2
+                end
+
+                return out
+            end
```

Comment:
Currently we are testing the non-function-call RHS at the same time as we test nested for-in loops. IMO, it would be better to have that as a separate test case that tests just the non-function-call case.

```
export function double_with_weird_rhs(xs: {any}): {integer}
    local out: {integer} = {}
    for i, x in iter as iterfn, xs as any, 0 as any do
        out[i] = x as integer * 2
    end
    return out
end
```